### PR TITLE
[GH-222] Data Reduction support for Unity

### PR DESCRIPTION
**Pull request by yong-huang**

### CODE REVIEW

File: storops/unity/resource/cg.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def modify_lun(self, lun, name=None, size=None, host_access=None,
             sp=sp,
             io_limit_policy=io_limit_policy,
             tiering_policy=tiering_policy,
-            is_compression=is_compression)
+            is_data_reduction_enabled=is_data_reduction_enabled)
```
Comment (Murray-LIANG) [inline storops/unity/resource/cg.py]: should pass in the `cli` to `prepare_lun_parameters`.

File: storops/unity/resource/lun.py [hunk 2]
```diff
@@ -37,7 +37,21 @@
 log = logging.getLogger(__name__)
 
 
-def prepare_lun_parameters(**kwargs):
+def prepare_lun_parameters(cli=None, **kwargs):
+    @version('<4.3')
+    def make_data_reduction_body(cli=None,
+                                 is_data_reduction_enabled=None):
+        if is_data_reduction_enabled is None:
```
Comment (Murray-LIANG) [inline storops/unity/resource/lun.py]: You could use `UnityClient.make_body` which will remove the None value key-value pair from the dict. Then there is no need to check `if is_data_reduction_enabled is None`.

File: storops/unity/resource/lun.py [hunk 3]
```diff
@@ -86,7 +105,7 @@ def get_nested_properties(cls):
     def create(cls, cli, name, pool, size, sp=None, host_access=None,
                is_thin=None, description=None, io_limit_policy=None,
                is_repl_dst=None, tiering_policy=None, snap_schedule=None,
-               is_compression=None):
+               is_data_reduction_enabled=None):
```
Comment (Murray-LIANG) [inline storops/unity/resource/lun.py]: You shouldn't change the API directly. It breaks the back-port capability.
You could append the new parameter after `is_compression`, and make these two parameters the same.

File: storops/unity/resource/lun.py [hunk 4]
```diff
@@ -191,33 +210,34 @@ def _compose_lun_parameter(cli, **kwargs):
...
+        lun_parameters = prepare_lun_parameters(cli, **kwargs)
         if lun_parameters:
             body['lunParameters'] = lun_parameters
         return body
 
     def modify(self, name=None, size=None, host_access=None,
                description=None, sp=None, io_limit_policy=None,
                is_repl_dst=None, tiering_policy=None, snap_schedule=None,
-               is_compression=None):
+               is_data_reduction_enabled=None):
```
Comment (Murray-LIANG) [inline storops/unity/resource/lun.py]: Ditto.

File: storops/unity/resource/lun.py [hunk 5]
```diff
@@ -191,33 +210,34 @@ def _compose_lun_parameter(cli, **kwargs):
...
-                                      io_limit_policy=io_limit_policy,
-                                      tiering_policy=tiering_policy,
-                                      is_compression=is_compression)
+            return self.cg.modify_lun(
+                self, name=name, size=size,
+                host_access=host_access,
+                description=description, sp=sp,
+                io_limit_policy=io_limit_policy,
+                tiering_policy=tiering_policy,
+                is_data_reduction_enabled=is_data_reduction_enabled)
```
Comment (Murray-LIANG) [inline storops/unity/resource/lun.py]: Ditto.

File: storops/unity/resource/pool.py [hunk 6]
```diff
@@ -101,16 +102,18 @@ def create_filesystem(self, nas_server, name, size,
     def create_lun(self, lun_name=None, size_gb=1, sp=None, host_access=None,
                    is_thin=None, description=None, tiering_policy=None,
                    is_repl_dst=None, snap_schedule=None, io_limit_policy=None,
-                   is_compression=None):
+                   is_data_reduction_enabled=None):
```
Comment (Murray-LIANG) [inline storops/unity/resource/pool.py]: Ditto.

File: storops/unity/resource/lun.py (hunk 3 above)
Comment (amagicsun) [reply storops/unity/resource/lun.py]: Was "is_compression" ever used somewhere? We didn't support it before, did we?

File: storops/unity/resource/cg.py (hunk 1 above)
Comment (yong-huang) [reply storops/unity/resource/cg.py]: Fixed.

File: storops/unity/resource/lun.py (hunk 2 above)
Comment (yong-huang) [reply storops/unity/resource/lun.py]: Done.

File: storops/unity/resource/lun.py (hunk 3 above)
Comment (yong-huang) [reply storops/unity/resource/lun.py]: Done, using the legacy parameter is_compression.

File: storops/unity/resource/lun.py (hunk 4 above)
Comment (yong-huang) [reply storops/unity/resource/lun.py]: Done.

File: storops/unity/resource/lun.py (hunk 5 above)
Comment (yong-huang) [reply storops/unity/resource/lun.py]: Done.

File: storops/unity/resource/pool.py (hunk 6 above)
Comment (yong-huang) [reply storops/unity/resource/pool.py]: Done.

File: storops/unity/resource/lun.py (hunk 3 above)
Comment (yong-huang) [reply storops/unity/resource/lun.py]: @amagicsun No, we didn't support it in Openstack cinder driver, but maybe someone use storops directly, and call this function in their own codes.

**Pull request merged**